subscriptions/logic/notifications.py

```diff
@@ -0,0 +1,171 @@
+            self._notifications[name] = _Notification(name, function, queryset, parameters, forget_after=forget_after)
+            return function
+        return decorator
+
+    @staticmethod
+    def _get_queryset(notification: _Notification) -> QuerySet:
+        filters = {}
+
+        queryset = notification.queryset.exclude(
+            subscription_notification_events__name=notification.name,
+        )
+        for parameter, value in notification.parameters.items():
+            # in the case of days since we do an aggregation.
+            # we do want to be sure it is the last item
+            if 'days_since_' in parameter:
+                parameter = parameter[len('days_since_'):]
+                since = now() - timedelta(days=value)
+                filters.update({
+                    f'{parameter}__gte': since,
+                    f'{parameter}__lte': since + timedelta(1)
+                })
+            else:
+                filters[parameter] = value
+        return queryset.filter(**filters)
+
+    @classmethod
+    def execute(cls, name: str) -> None:
+        """ Fetch all the missing notifications. """
+
+        # to ensure the same entry has been generated only once
+        with HardDBLock(cls.__name__, name):
+            notification = cls._notifications[name]
+            # eventually some notification might be designed to be sent after enough time passed.
+            # for example, see this scenario
+            # --- unsubscribe -> coupon -> subscribe -> unsubscribe -> coupon ? ---
+            #  THe coupon might be sent if enough time has passed since last time.
+            if notification.forget_after is not None:
+                # TODO: There is a reason why not to delete these?
+                SubscriptionNotificationEvent.objects.filter(
+                    name=notification.name,
+                    created__lte=now() - timedelta(notification.forget_after)
+                ).delete()
+            queryset = cls._get_queryset(notification)
+            events = [
+                SubscriptionNotificationEvent(user=user, name=name)
+                for user in queryset
+            ]
+            SubscriptionNotificationEvent.objects.bulk_create(
+                events
+            )
+        for event in events:
+            cls._execute_event(event, notification)
+
+    @classmethod
+    def _execute_event(cls, event: SubscriptionNotificationEvent, notification: _Notification):
+        with cls._event_handler(event):
+            notification.function(event.user)
+
+    @classmethod
+    @contextmanager
+    def _event_handler(cls, event: SubscriptionNotificationEvent):
+
+        event.status = event.Status.ONGOING
+        event.save()
+        try:
+            yield
+            event.status = event.Status.COMPLETED
+            event.save()
+        except Exception:
+            event.status = event.Status.ERROR
+            event.save()
+            raise
```

Comment:
TODO: this should not raise, I did not write an appropriate test so I kept it like this, but probably it should rather log the error